### ISSUE TITLE
Use columns for properties when dumping collections? 

### ISSUE THREAD

**Issue by wilka** [MoaidHathot/Dumpify#30]
One of the handy things with LinqPad `.Dump()` is when you use it on a collection, you get a table with columns for each property. e.g. in LinqPad the code:
```csharp
void Main()
{
	var people = new[] 
	{
		new Person("Marie", "Curie"),   
		new Person("Albert", "Einstein"),
		new Person("Ada", "Lovelace"),
		new Person("Cleopatra", "Philopator"),
		new Person("Nikola", "Tesla")
	};

	people.Dump();
}

record Person(string FirstName, string LastName);
```

gives:
![image](https://github.com/MoaidHathot/Dumpify/assets/1788792/b912b557-50c3-4b22-8999-6bbdcb8f7f0b)

but with Dumpify, it shows:
![image](https://github.com/MoaidHathot/Dumpify/assets/1788792/ecbb86d1-1711-4ce6-a787-c55fb24da67b)

So I'd like to be able to have the more concise column-for-property format with collections.

**Comment by MoaidHathot** [comment]
Hi @wilka, this feature is already supported but it is turned-off be default. It can be turned-on either globally for all dumps or per-dump:

Globally:
`DumpConfig.Default.TableConfig.ShowRowSeparators = true;`
 
Per dump:
`people.Dump(tableConfig: new TableConfig { ShowRowSeparators = true });`

Turning this option on results in the following output:
![image](https://github.com/MoaidHathot/Dumpify/assets/8770486/d94d71bd-8278-4c4b-9f1d-cc9dc27b43c8)

**Comment by wilka** [comment]
If it makes any difference, I'm not _hugely_ concerned about the terminal width limits. I tend to run my Windows Terminal at full screen, and I have a large display.  But I realise that won't be the case for everyone.

I've been using Dumpify as a stepping stone from taking some LinqPad prototype code, and then turning it into a console app.

**Comment by MoaidHathot** [comment]
@wilka, is this similar to what you were looking for? (Ignore the circular reference bug)


![Image](https://github.com/user-attachments/assets/9d9e4699-8cec-4570-bbbf-0aa8747ed618)

Instead of 

![Image](https://github.com/user-attachments/assets/3d7b5737-366f-4272-b5a5-4d8db7d90fa8)

**Comment by wilka** [comment]
Yup, that is what I had in mind. The difference is more useful when you have more) records, so you only have a single header row shared for multiple items.

**Comment by Tdue21** [comment]
I'd love this!!

**Comment by NickNameIsNull** [comment]
Has this been implemented in the current version? If yes, how should I configure it to achieve such a display? If not, in which version is it expected to be implemented?

**Comment by MoaidHathot** [comment]
Hi @NickNameIsNull, sorry for the delay in response.
It will be part of the next version, currently testing.

**Comment by MoaidHathot** [comment]
I’ve implemented this feature and merged it into the main branch. It hasn’t been pushed to NuGet yet, as I’d like to gather more feedback first.
I’d really appreciate your thoughts. You can find examples in the Playground project within the repository.

The engine now supports:
- Defining rules per type
- Hybrid approaches
- Dynamic, user-configured layout selection at runtime

@wilka @Tdue21 @NickNameIsNull 

Examples:

<img width="299" height="331" alt="Image" src="https://github.com/user-attachments/assets/2bfce592-3082-400d-9aa2-71ce184462ef" />

<img width="357" height="205" alt="Image" src="https://github.com/user-attachments/assets/8abf66b0-f1bf-4eb5-8f21-318a7f7d75fe" />

<img width="439" height="276" alt="Image" src="https://github.com/user-attachments/assets/6f35f7ad-41b8-4eb2-893e-96456f33da9a" />

<img width="380" height="237" alt="Image" src="https://github.com/user-attachments/assets/2fc41b32-044b-484f-aa14-6e1053bf4edb" />

<img width="550" height="638" alt="Image" src="https://github.com/user-attachments/assets/bb0066ab-448e-4112-a02b-88d90de97f95" />

<img width="611" height="774" alt="Image" src="https://github.com/user-attachments/assets/67d3b005-d56f-45a1-a59b-b1d7a618ca06" />

<img width="526" height="448" alt="Image" src="https://github.com/user-attachments/assets/87ef6ada-520e-4800-b5b0-0e247da3ba9f" />

**Comment by ThomasDue** [comment]
It *looks* promising, but it is not easy to play around with. I understand the not wanting to release until you've had more feedback. I'd suggest making a pre-release of the nuget then. 
The alternative:
1. Remove the nuget from my project. 
2. Download the repo 
3. link into my project instead of just updating the nuget

**Comment by MoaidHathot** [comment]
@ThomasDue, good point!
I've published a pre-release version 0.7.1-beta.1 
https://www.nuget.org/packages/Dumpify/0.7.1-beta.1 
Example configuration can be seen and used here:
https://github.com/MoaidHathot/Dumpify/blob/main/src/Dumpify.Playground/Program.cs

@wilka @Tdue21 @NickNameIsNull

**Comment by ThomasDue** [comment]
It looks promising. Horizontal Layout does not seem to honor `MembersConfig.MemberFilter` though. 

I have tried various variations of this: 
```csharp
var limitColumns = new MembersConfig
{
    MemberFilter = ctx => ctx.Member.Name == "ElectricCarAccount"
                       || ctx.Member.Name == "StartTime"
                       || ctx.Member.Name == "ModifiedTime"
                       || ctx.Member.Name == "PriceInclVat"
                       || ctx.Member.Name == "AmountInKw"
};
result .Dump(members: limitColumns, tableConfig: new() { TableLayout = TableLayout.Horizontal });
```
But it outputs all public properties no matter what I do. I *can* make a `.Select(x => new { ... })` on the result set *before* the `.Dump()`, but the built-in filtering does not work.

**Comment by MoaidHathot** [comment]
Thanks @ThomasDue, that’s a great catch.

It highlights a challenge in how horizontal properties are handled for collections. A collection may be statically typed as` Base[]`, while at runtime it can contain multiple derived types, for example: `Base[] b = [new Derived1(), new Derived2()]`.

That raises the question of which columns should be displayed.

1. Show only the columns defined on `Base`, regardless of any additional properties on derived types. This is simpler to implement, but it could be misleading or hide useful information.
1. Merge the columns from all encountered types and display an empty value indicator or `N/A` when a given object does not expose a particular column. This is more flexible, but also more complex.
1. Add a configuration option that lets users decide how this scenario should be handled. Most complex.

I'm leaning towards option 3, but even with that option, we need to decide upon a default.

**Comment by ThomasDue** [comment]
I see your point. To be honest I don't have a strong option either way (right now, that might change). As a general rule I'd expect the table to show the properties of the base type. I.e. your option 1. I think this is the right way, with some kind of option to override, i.e. like option 3. 

In this particular case, I'd just like for the member filter option to be honored ;)
